src/row_codec.rs

```diff
@@ -1,182 +1,230 @@
-use crate::db_state::RowFeature;
 use crate::error::SlateDBError;
-use crate::types::{KeyValueDeletable, RowAttributes, ValueDeletable};
+use crate::types::{RowEntry, ValueDeletable};
 use bitflags::bitflags;
 use bytes::{Buf, BufMut, Bytes, BytesMut};
 
 bitflags! {
+    #[derive(Debug, Clone, PartialEq, Default)]
     pub(crate) struct RowFlags: u8 {
         const Tombstone = 0b00000001;
+        const HAS_EXPIRE_TS = 0b00000010;
+        const HAS_CREATE_TS = 0b00000100;
     }
 }
 
-const NO_EXPIRE_TS: i64 = i64::MIN;
-
 /// Encodes key and value using the binary codec for SlateDB row representation
-/// using the `v0` encoding scheme.
+/// using the `v1` encoding scheme.
 ///
-/// The `v0` codec for the key is (for non-tombstones):
+/// The `v1` codec for the key is (for non-tombstones):
 ///
 /// ```txt
-///  |--------------------------------------------------------------------------|
-///  |       u16      |      u16       |  var        | var  |    u32    |  var  |
-///  |----------------|----------------|-------------|------|-----------|-------|
-///  | key_prefix_len | key_suffix_len |  key_suffix | meta | value_len | value |
-///  |--------------------------------------------------------------------------|
+///  |---------------------------------------------------------------------------------------------------------------------|
+///  |       u16      |      u16       |  var        | u64     | u8        | i64       | i64       | u32       |  var      |
+///  |----------------|----------------|-------------|---------|-----------|-----------|-----------|-----------|-----------|
```

Comment:
I think in future PRs, there would be changes to read a key with `seq <= s` to do snapshot reads. 
1. In which scenarios would we have multiple values for the same key, with the same seq in the same block? If yes, what would be the sort order for the seq? Given we are looking for largest seq less than given value, would desc be better?
2. Right now, we exit the moment we find the key in the lookup order (i.e. memtable, l0, SR1, SR2...). With seq, would we still be able to do it?  I think yes, because seq for a key in SR2 must be less than its seq in SR1 or L0.

Reply:
> In which scenarios would we have multiple values for the same key, with the same seq in the same block?

in my understanding, we might have change one key within a short time period, like, we may update "key1" 100 times within 1 second, each update will generate a new sequence number, and all these values co-locates in the memtable, and highly likely to be dumped in a same block.

> If yes, what would be the sort order for the seq?

yes, sorting helps here, we can always put the rows with higher seq numbers up front, which would helps scanning within these multi version KVs.

> Right now, we exit the moment we find the key in the lookup order (i.e. memtable, l0, SR1, SR2...). With seq, would we still be able to do it? I think yes, because seq for a key in SR2 must be less than its seq in SR1 or L0.

+1, this is an interesting aspect i haven't thought before, yes, the sequence number aligns well with the levels of SSTs. the keys in higher level's SSTs (like SR1) should have a higher sequence number than the lower level's SSTs (like SR3).

Reply:
by the way we'll have a similar thing when we support "as of" queries, we'll need to also consider sorting by create_ts (descending) because we'll allow multiple copies of a key (I think most likely is to allow multiple copies within a certain window of time and expire when the diff between most recent and oldest is some limit)